src/box/applier.cc

```diff
@@ -369,6 +369,189 @@ applier_connection_init(struct iostream *io, const struct uri *uri,
+			      struct ibuf *ibuf, struct ballot *ballot,
+			      int *decoded_fields)
+{
+	RegionGuard guard(&fiber()->gc);
+	xrow_encode_watch_key(row, internal_ballot_key, IPROTO_WATCH);
+	coio_write_xrow(io, row);
+	coio_read_xrow(io, ibuf, row);
+	if (row->type != IPROTO_EVENT) {
+		xrow_decode_error_xc(row);
+	}
+	struct watch_request watch_request;
+	xrow_decode_watch_xc(row, &watch_request);
+	if (watch_request.data == NULL)
+		tnt_raise(ClientError, ER_NO_SUCH_EVENT, internal_ballot_key);
+	xrow_decode_ballot_event_xc(&watch_request, ballot, decoded_fields);
+}
+
+/** Get applier ballot via the newest of available methods. */
+static void
+applier_get_ballot(struct applier *applier)
+{
+	struct xrow_header row;
+	struct iostream io;
+	struct iostream_ctx io_ctx;
+	struct ibuf ibuf;
+	iostream_ctx_clear(&io_ctx);
+	if (iostream_ctx_create(&io_ctx, IOSTREAM_CLIENT, &applier->uri) != 0)
+		diag_raise();
+	iostream_clear(&io);
+	ibuf_create(&ibuf, &cord()->slabc, 1024);
+	auto guard = make_scoped_guard([&] {
+		if (iostream_is_initialized(&io))
+			iostream_close(&io);
+		iostream_ctx_destroy(&io_ctx);
+		ibuf_destroy(&ibuf);
+	});
+
+	struct greeting greeting;
+
+	applier->addr_len = sizeof(applier->addrstorage);
+	applier_connection_init(&io, &applier->uri, &applier->addr,
+				&applier->addr_len, &io_ctx, &greeting);
+	try {
+		while (true) {
+			int decoded_fields = 0;
+			applier_get_ballot_from_event(&io, &row, &ibuf,
+						      &applier->ballot,
+						      &decoded_fields);
+			/*
+			 * Some events received right after master start
+			 * might be empty. Ignore those.
+			 */
+			if (decoded_fields > 0)
+				applier_run_ballot_triggers(applier, true);
+		}
+	} catch (ClientError *e) {
+		if (e->errcode() != ER_UNKNOWN_REQUEST_TYPE &&
+		    e->errcode() != ER_NO_SUCH_EVENT)
+			throw;
+		applier_get_ballot_from_vote(&io, &row, &ibuf,
+					     &applier->ballot);
+		applier_run_ballot_triggers(applier, true);
+	}
+}
+
+static int
+applier_ballot_watcher_f(va_list ap)
+{
+	(void)ap;
+	struct applier *applier = (struct applier *)fiber()->f_arg;
+	while (true) {
+		try {
+			applier_get_ballot(applier);
+			return 0;
+		} catch (TimedOut *) {
+			/* debug */diag_log();
+			fiber_sleep(replication_reconnect_interval());
+			diag_clear(diag_get());
+		} catch (Exception *) {
+			diag_log();
+			applier_run_ballot_triggers(applier, false);
+			diag_clear(diag_get());
+			applier->ballot_watcher = NULL;
+			return 0;
+		}
+	}
+	return 0;
+}
+
+/** Trigger data for waiting for the first ballot update. */
+struct applier_ballot_data {
+	/** Diagnostics set in case of error. */
+	struct diag diag;
+	/** The fiber waiting for the ballot update. */
+	struct fiber *fiber;
+	/** Whether the ballot was updated. */
+	bool done;
+};
+
+static int
+applier_on_first_ballot_update_f(struct trigger *trigger, void *event)
+{
+	struct applier_ballot_data *data =
+		(struct applier_ballot_data *)trigger->data;
+	bool *success = (bool *)event;
```

Comment:
Can't this be rewritten as `bool success = *(bool *)event`? This way it will be easier to use afterwards.

Reply:
Sure, thanks for the suggestion!